src/extension.ts

```diff
@@ -136,6 +136,9 @@ export async function activate(ctx: vscode.ExtensionContext) {
   whenFileExists(ctx, compileCommandsFile, async () => {
     const conf = vscode.workspace.getConfiguration("C_Cpp");
     conf.update("default.compileCommands", compileCommandsFile, vscode.ConfigurationTarget.Workspace);
+
+    const conf2 = vscode.workspace.getConfiguration("clangd");
+    conf2.update("arguments", [`--compile-commands-dir=${buildDir}`], vscode.ConfigurationTarget.Workspace);
```

Comment:
The downside is we override user settings that could be defining other stuff than the compile commands. Maybe we should wait for clangd to add a dedicated setting for this?

BTW, if someone knows a better trick than a workspace setting for communicating the path to compile_commands.json to other extensions, that would be great. One big downside of this approach is it modifies `.vscode/settings.json` in user's source tree, and some projects does track that file in git...

Reply:
> Maybe we should wait for clangd to add a dedicated setting for this?

The PR for vscode-clangd is currently more than a month old, with zero comments so far. And it's not even that complicated... Currently I can't see it moving forward in a reasonable timeframe.

> BTW, if someone knows a better trick than a workspace setting for communicating the path to compile_commands.json to other extensions, that would be great.

In theory, extensions should be able to expose their internal API to other extensions by returning data from the activation function.

https://code.visualstudio.com/api/references/vscode-api#extensions

> One big downside of this approach is it modifies `.vscode/settings.json` in user's source tree, and some projects does track that file in git...

True.

Reply:
> compile_commands.json to other extensions, that would be great. One big downside of this approach is it modifies .vscode/settings.json in user's source tree, and some projects does track that file in git...

I actually hate this about the Meson extension. For instance, I have `"C_Cpp.intelliSenseEngine": "disabled",`, so I don't even use the compilation database for the extension. We should check this setting to not edit the file.

I've also thought about just adding a setting to this extension whether or not to auto setup other extensions. It is really freaking annoying!

Reply:
@wolfpld I think we could be smarter here. If we can read the current setting, we can then check if the compile commands dir is already set. If not set, extend the current setting with the arguments to set the compile commands dir.

Reply:
> I've also thought about just adding a setting to this extension whether or not to auto setup other extensions. It is really freaking annoying!

I made this configurable.

Reply:
> @wolfpld I think we could be smarter here. If we can read the current setting, we can then check if the compile commands dir is already set. If not set, extend the current setting with the arguments to set the compile commands dir.

Something like this?

```javascript
      let args = undefined;
      const wanted = `--compile-commands-dir=${buildDir}`;
      const current = conf.get("clangd.arguments");
      if (current && Array.isArray(current)) {
        // if wanted is already in current, leave args undefined to prevent a config update
        if (!current.includes(wanted)) {
          const found = current.findIndex((arg) => arg.startsWith("--compile-commands-dir="));
          if (found !== -1) {
            current[found] = wanted;
            args = current;
          } else {
            args = [...current, wanted];
          }
        }
      } else {
        args = [wanted];
      }

      if (args) {
        conf
          .update("clangd.arguments", args, vscode.ConfigurationTarget.Workspace)
          .then(
            () => vscode.commands.executeCommand("clangd.restart"),
            () => {},
          );
      }
```

Reply:
> FWIW, the clangd extension seems to do a recursive search for compilation databases.

It won't pick up the database if I don't change the default `builddir` to `build` in the meson extension (without this PR).

I also had some very unfun experience trying to figure out why the include path changes I made in `meson.build` were not picked up by clangd, while the project compiled just fine. It turned out that I had two build dirs, `build` and `builddir`. Clangd was reading `build` while the meson extension was updating `builddir`.

Reply:
> Perhaps. I would need to test with my setup.

Ok, I pushed it. Note that the `clangd.arguments` getter is not scoped to the workspace, so it will pick up user settings if any are set, and the workspace settings are empty.